source/opt/graphics_robust_access_pass.cpp

```diff
@@ -0,0 +1,984 @@
+//
+//        - Otherwise, the pointer type fully encodes the range of valid
+//          addresses. In particular, the type of a pointer to an aggregate
+//          value fully encodes the range of indices when indexing into
+//          that aggregate.
+//
+//    - The pointer is the result of an access chain instruction.  We clamp
+//      indices contributing to address calculations.  As noted above, the
+//      valid ranges are either bound by the length of a runtime array, or
+//      by the type of the base pointer.  The length of a runtime array is
+//      the result of an OpArrayLength instruction acting on the poitner of
+//      the containing structure as noted above.
+//
+//    - TODO(dneto): OpImageTexelPointer:
+//      - Clamp coordinate to the image size returned by OpImageQuerySize
+//      - If multi-sampled, clamp the sample index to the count returned by
+//        OpImageQuerySamples.
+//      - If not multi-sampled, set the sample index to 0.
+//
+//  - Rely on the external validator to check that pointers are only
+//    used by the instructions as above.
+//
+//  - Handles OpTypeRuntimeArray
+//    Track pointer back to original resource (pointer to struct), so we can
+//    query the runtime array size.
+//
+
+#include "graphics_robust_access_pass.h"
+
+#include <algorithm>
+#include <cstring>
+#include <functional>
+#include <initializer_list>
+#include <utility>
+
+#include "constants.h"
+#include "def_use_manager.h"
+#include "function.h"
+#include "ir_context.h"
+#include "module.h"
+#include "pass.h"
+#include "source/diagnostic.h"
+#include "source/util/make_unique.h"
+#include "spirv-tools/libspirv.h"
+#include "spirv/unified1/GLSL.std.450.h"
+#include "spirv/unified1/spirv.h"
+#include "type_manager.h"
+#include "types.h"
+
+namespace spvtools {
+namespace opt {
+
+using opt::BasicBlock;
+using opt::Instruction;
+using opt::Operand;
+using spvtools::MakeUnique;
+
+GraphicsRobustAccessPass::GraphicsRobustAccessPass() : module_status_() {}
+
+Pass::Status GraphicsRobustAccessPass::Process() {
+  module_status_ = PerModuleState();
+
+  ProcessCurrentModule();
+
+  auto result = module_status_.failed
+                    ? Status::Failure
+                    : (module_status_.modified ? Status::SuccessWithChange
+                                               : Status::SuccessWithoutChange);
+
+  return result;
+}
+
+spvtools::DiagnosticStream GraphicsRobustAccessPass::Fail() {
+  module_status_.failed = true;
+  // We don't really have a position, and we'll ignore the result.
+  return std::move(
+      spvtools::DiagnosticStream({}, consumer(), "", SPV_ERROR_INVALID_BINARY)
+      << name() << ": ");
+}
+
+spv_result_t GraphicsRobustAccessPass::IsCompatibleModule() {
+  auto* feature_mgr = context()->get_feature_mgr();
+  if (!feature_mgr->HasCapability(SpvCapabilityShader))
+    return Fail() << "Can only process Shader modules";
+  if (feature_mgr->HasCapability(SpvCapabilityVariablePointers))
```

Comment:
? 
The first line ensures we must have shader capability
This line ensures we *don't* have variable pointers.
Plus, I want a specific error message to match the condition that was found. It would be confusing for most people to have VariablePointers in the module but the the error beeing "hey you've turned on VariablePointersStorageBuffer"....